fix(kotlin) number literals

flofriday opened the pull request:
Previously the parser for Kotlin reused the Java number parsing but there are significant differences in how the two languages express numbers causing false-negatives and false-positives.

<!--- Provide a general summary of your changes in the Title above -->

<!-- Please link to a related issue below. -->
<!-- ie, `Resolves #1234`, etc... so GitHub can magically link it -->
Resolves #4305

### Changes
Implemented a custom parser for Kotlin number literals since they are different enough from Java literals to justify their own parser.

My hunch is that the orginal implementor misinterpreted the [release notes from Kotlin 1.1](https://kotlinlang.org/docs/reference/whatsnew11.html#underscores-in-numeric-literals) which they referenced since this only applies to how seperators in Kotlin follow the same rules as Java 8 and not how all the number parsing is identical to Java 8.
Moreover, in the 9 years since Kotlin 1.1 the language has changed a lot and for example back then unsigned number literals surely weren't a thing.

### Checklist
- [x] Added markup tests, or they don't apply here because...
- [x] Updated the changelog at `CHANGES.md`

flofriday committed:
fix(kotlin) number literals

Previously kotlin resused the java number parsing but there are
significant differnces in how the two languages express numbers
causing false-negatives and false-positives.

joshgoebel commented on src/languages/kotlin.js:
```diff
@@ -100,10 +98,31 @@ export default function(hljs) {
-  // so the code below is copied from java.js
-  const KOTLIN_NUMBER_MODE = NUMERIC;
+  const decimalDigits = '[0-9](_*[0-9])*';
+  const frac = `\\.(${decimalDigits})`;
```
```suggestion
  const fractional = `\\.(${decimalDigits})`;
```

joshgoebel commented on src/languages/kotlin.js:
```diff
@@ -100,10 +98,31 @@ export default function(hljs) {
+  const decimalDigits = '[0-9](_*[0-9])*';
+  const frac = `\\.(${decimalDigits})`;
+  const KOTLIN_NUMBER_MODE = {
+    className: 'number',
```
```suggestion
    scope: 'number',
```

joshgoebel commented on src/languages/kotlin.js:
```diff
@@ -100,10 +98,31 @@ export default function(hljs) {
+    variants: [
+      // DecimalFloatingPointLiteral
+      // including ExponentPart
+      { begin: `(\\b(${decimalDigits})((${frac})|\\.)?|(${frac}))` +
```
```suggestion
      { match: `(\\b(${decimalDigits})((${frac})|\\.)?|(${frac}))` +
```

Prefer match for all these if there is no `end`.

flofriday replied:
thanks :)

flofriday committed:
fix(kotlin) number literals

Previously kotlin resused the java number parsing but there are
significant differnces in how the two languages express numbers
causing false-negatives and false-positives.

flofriday commented:
Thanks for the suggestions, I implemented them all :)

joshgoebel commented:
Did you? I'm not seeing the changes.

flofriday commented:
Uff, that's embarrassing, no I forgot to push 🙈

joshgoebel commented on src/languages/kotlin.js:
```diff
@@ -100,10 +98,31 @@ export default function(hljs) {
+      { match: `\\b0[xX]([0-9a-fA-F](_*[0-9a-fA-F])*)[uU]?L?\\b` },
+  
+      // BinaryIntegerLiteral
+      { match: '\\b0[bB][01](_*[01])*[uU]?L?\\b' },
```
Can we make all these actual regex to avoid the double \\ escaping?

joshgoebel commented on src/languages/kotlin.js:
```diff
@@ -100,10 +98,31 @@ export default function(hljs) {
+      // BinaryIntegerLiteral
+      { match: '\\b0[bB][01](_*[01])*[uU]?L?\\b' },
+    ],
+    relevance: 0
```
Don't need relevance anymore.

joshgoebel commented:
Please lets go backwards (before rebase), the cedar doc change should NOT be a part of this PR.

flofriday commented:
> [..] the cedar doc change should NOT be a part of this PR.

Totally agree, I cannot find any cedar changes though, maybe it was in an older commit by accident which resolved itself on its own.

I now also included all the other changes.

joshgoebel commented on src/languages/kotlin.js:
```diff
@@ -100,10 +98,30 @@ export default function(hljs) {
+    variants: [
+      // DecimalFloatingPointLiteral
+      // including ExponentPart
+      { match: new RegExp(`(\\b(${decimalDigits})((${fractional})|\\.)?|(${fractional}))` +
```
`new RegExp(...)` isn’t needed here — hljs accepts string patterns the same as regex literals. Same style as `src/languages/lib/java.js`:

```js
{ match: `(\\b(${decimalDigits})((${fractional})|\\.)?|(${fractional}))` +
  `[eE][+-]?(${decimalDigits})[fF]?\\b` },
{ match: `\\b(${decimalDigits})((${fractional})[fF]?\\b|\\.([fF]\\b)?)` },
{ match: `(${fractional})[fF]?\\b` },
{ match: `\\b(${decimalDigits})[fF]\\b` },
```

(Applies to the four float variants below too.)

joshgoebel commented on src/languages/kotlin.js:
```diff
@@ -100,10 +98,30 @@ export default function(hljs) {
-  // According to the doc above, the number mode of kotlin is the same as java 8,
-  // so the code below is copied from java.js
-  const KOTLIN_NUMBER_MODE = NUMERIC;
+  const decimalDigits = '[0-9](_*[0-9])*';
```
Could we get a markup test for `_`... and truly is `5______9` a valid number?

flofriday replied:
Yes `5______9` is really a valid number, here is a [Kotlin Playground example](https://pl.kotl.in/3WNqc6Hmw).
